src/CodeGen_GPU_Dev.cpp

```diff
@@ -162,25 +167,56 @@ void CodeGen_GPU_C::visit(const Shuffle *op) {
         std::string src = vecs[0];
         std::ostringstream rhs;
         std::string storage_name = unique_name('_');
-        if (vector_declaration_style == VectorDeclarationStyle::OpenCLSyntax) {
+        switch (vector_declaration_style) {
+        case VectorDeclarationStyle::OpenCLSyntax:
             rhs << "(" << print_type(op->type) << ")(";
-        } else if (vector_declaration_style == VectorDeclarationStyle::WGSLSyntax) {
+            break;
+        case VectorDeclarationStyle::WGSLSyntax:
             rhs << print_type(op->type) << "(";
-        } else {
+            break;
+        case VectorDeclarationStyle::CLikeSyntax:
             rhs << "{";
+            break;
         }
+        int elem_num = 0;
         for (int i : op->indices) {
-            rhs << vecs[i];
-            if (i < (int)(op->indices.size() - 1)) {
+            size_t vector_idx;
+            int lane_idx = -1;
+            for (vector_idx = 0; vector_idx < op->vectors.size(); ++vector_idx) {
+                if (i >= vector_first_index[vector_idx] && i < vector_first_index[vector_idx] + op->vectors[vector_idx].type().lanes()) {
+                    lane_idx = i - vector_first_index[vector_idx];
+                    break;
+                }
+            }
+            internal_assert(lane_idx != -1) << "Shuffle lane index not found: i=" << i;
+            internal_assert(vector_idx < op->vectors.size()) << "Shuffle vector index not found: i=" << i << ", lane=" << lane_idx;
+            rhs << vecs[vector_idx];
+            if (op->vectors[vector_idx].type().lanes() > 1) {
```

Comment:
Won’t these vectors always have lanes > 1?

Reply:
No, that was the assumption before this PR. Everything was assumed to be scalar. Shuffling actual vectors was broken.